File: RaveX-ravex146/src/main/java/ravex/modules/world/AutoNameTag.java
package ravex.modules.world;

import ravex.modules.Category;
import ravex.modules.Module;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.entity.decoration.ArmorStand;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.InteractionHand;

public class AutoNameTag extends Module {
    public static final AutoNameTag INSTANCE = new AutoNameTag();

    private AutoNameTag() {
        super("AutoNameTag", Category.WORLD);
    }

    @Override
    public void onTick() {
        Minecraft mc = Minecraft.getInstance();
        LocalPlayer p = mc.player;
        if (p == null || mc.level == null || mc.gameMode == null) return;


        int tagSlot = -1;
        ItemStack tagStack = null;
        for (int i = 0; i < 9; i++) {
            ItemStack stack = p.getInventory().getItem(i);
            if (stack.is(Items.NAME_TAG) && stack.getCustomName() != null) {
                tagSlot = i;
                tagStack = stack;
                break;
            }
        }

        if (tagSlot == -1 || tagStack == null) return;
        String tagName = tagStack.getHoverName().getString();


        LivingEntity target = null;
        double closestDist = 4.5;
        for (var entity : mc.level.entitiesForRendering()) {
            if (entity instanceof LivingEntity living && !(entity instanceof Player) && !(entity instanceof ArmorStand)) {
                if (living.isAlive()) {

                    if (living.getCustomName() != null && tagName.equals(living.getCustomName().getString())) {
                        continue;
                    }
                    double dist = p.distanceTo(living);
                    if (dist < closestDist) {
                        closestDist = dist;
                        target = living;
                    }
                }
            }
        }

        if (target == null) return;


        int prevSlot = p.getInventory().getSelectedSlot();
        p.getInventory().setSelectedSlot(tagSlot);

        mc.gameMode.interact(p, target, InteractionHand.MAIN_HAND);
        p.swing(InteractionHand.MAIN_HAND);

        if (tagSlot != prevSlot) {
            p.getInventory().setSelectedSlot(prevSlot);
        }
    }
}

File: src/main/kotlin/volthack/modules/world/AutoNameTag.kt
package volthack.modules.world

import net.minecraft.client.Minecraft
import net.minecraft.world.InteractionHand
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.item.Items
import volthack.event.EventBus
import volthack.event.TickEvent
import volthack.setting.Category
import volthack.setting.Module

object AutoNameTag : Module("AutoNameTag", "Automatically applies Name Tags to nearby living entities that do not already have a custom name", Category.WORLD) {
    private val reachRange by float("Range", 4.5f, 2.0f, 6.0f, 0.1f)
    private val useDelay by int("Delay Ticks", 10, 2, 40)

    private var cooldown = 0

    init {
        EventBus.listen<TickEvent> { onTick() }
    }

    private fun onTick() {
        if (!enabled) return

        if (cooldown > 0) {
            cooldown--
            return
        }

        val mc = Minecraft.getInstance()
        val player = mc.player ?: return
        val world = mc.level ?: return

        // 1. Locate Name Tag in hotbar
        val tagSlot = (0..8).firstOrNull { player.inventory.getItem(it).item == Items.NAME_TAG } ?: return

        // 2. Scan for nearby living entities without a custom name
        val target = world.entitiesForRendering()
            .filterIsInstance<LivingEntity>()
            .filter { it != player && it.isAlive }
            .filter { player.distanceToSqr(it) <= (reachRange * reachRange).toDouble() }
            .firstOrNull { !it.hasCustomName() }

        if (target != null) {
            // Swap to Name Tag slot
            val prevSlot = player.inventory.selectedSlot
            player.inventory.selectedSlot = tagSlot

            // Interact with entity
            mc.gameMode?.interact(player, target, InteractionHand.MAIN_HAND)
            player.swing(InteractionHand.MAIN_HAND)

            // Restore slot
            player.inventory.selectedSlot = prevSlot

            cooldown = useDelay
        }
    }

    override fun onDisable() {
        cooldown = 0
    }
}
